rust-install/src/mock/clitools.rs

```diff
@@ -0,0 +1,392 @@
+    let ref url = Url::from_file_path(config.distdir.path()).unwrap();
+    for channel in &["nightly", "beta", "stable"] {
+        change_channel_date(url, channel, date);
+    }
+}
+
+pub fn expect_ok(config: &Config, args: &[&str]) {
+    expect_stdout_ok(config, args, "");
+}
+
+pub fn expect_err(config: &Config, args: &[&str], expected: &str) {
+    let out = run(config, args[0], &args[1..], &[]);
+    println!("out.ok: {}", out.ok);
+    println!("out.stdout:\n\n{}", out.stdout);
+    println!("out.stderr:\n\n{}", out.stderr);
+    println!("expected: {}", expected);
+    let args = format!("{:?}", args);
+    assert!(!out.ok, args);
+    assert!(out.stderr.contains(expected), args);
+}
+
+pub fn expect_stdout_ok(config: &Config, args: &[&str], expected: &str) {
+    let out = run(config, args[0], &args[1..], &[]);
+    println!("out.ok: {}", out.ok);
+    println!("out.stdout:\n\n{}", out.stdout);
+    println!("out.stderr:\n\n{}", out.stderr);
+    println!("expected: {}", expected);
+    let args = format!("{:?}", args);
+    assert!(out.ok, args);
+    assert!(out.stdout.contains(expected), args);
+}
+
+pub fn expect_stderr_ok(config: &Config, args: &[&str], expected: &str) {
+    let out = run(config, args[0], &args[1..], &[]);
+    println!("out.ok: {}", out.ok);
+    println!("out.stdout:\n\n{}", out.stdout);
+    println!("out.stderr:\n\n{}", out.stderr);
+    println!("expected: {}", expected);
+    let args = format!("{:?}", args);
+    assert!(out.ok, args);
+    assert!(out.stderr.contains(expected), args);
+}
+
+#[derive(Debug)]
+pub struct SanitizedOutput {
+    pub ok: bool,
+    pub stdout: String,
+    pub stderr: String,
+}
+
+pub fn run(config: &Config, name: &str, args: &[&str], env: &[(&str, &str)]) -> SanitizedOutput {
+    let exe_path = config.exedir.path().join(format!("{}{}", name, EXE_SUFFIX));
+    let mut cmd = Command::new(exe_path);
+    cmd.args(args);
+    cmd.env("MULTIRUST_HOME", config.homedir.path().to_string_lossy().to_string());
+    cmd.env("MULTIRUST_DIST_ROOT", format!("file://{}", config.distdir.path().join("dist").to_string_lossy()));
+    for env in env {
+        cmd.env(env.0, env.1);
+    }
+    let out = cmd.output().unwrap();
+
+    SanitizedOutput {
+        ok: out.status.success(),
+        stdout: String::from_utf8(out.stdout).unwrap(),
+        stderr: String::from_utf8(out.stderr).unwrap(),
+    }
+}
+
+/// Holds a process wide mutex while working in another directory.
+/// Use this any time you are invoking 'multurust override', since
+/// that captures the current directory from the environment, and
+/// will race without the mutex.
+pub fn change_dir(path: &Path, f: &Fn()) {
+    lazy_static! {
+        static ref LOCK: Mutex<()> = Mutex::new(());
+    }
+    let _g = LOCK.lock();
+    let cwd = env::current_dir().unwrap();
+    env::set_current_dir(path).unwrap();
+    let _g = scopeguard::guard((), move |_| env::set_current_dir(&cwd).unwrap());
```

Comment:
I'd personally feel that defining a shim here with a `Drop` implementation would be easier than pulling in a whole crate to do this job, but either way's probably fine.

If this function stays it may also be best to return an RAII guard instead of taking a closure (helps indentation, etc)